[Serve] Implement 'sky serve apply' for services

jusstinn opened the pull request:
<!-- Describe the changes in this PR -->
This PR implements a unified `sky serve apply` interface that supports both services and worker pools. Previously, this command raised an assertion error when used with standard services.

**Key changes:**
* **Upsert Logic:** Implemented "Update or Insert" logic in `sky.serve.client.impl.apply`.
    * Checks if the service exists using `status()`.
    * If the service exists, it triggers an `update()`.
    * If the service does not exist (or the controller is down/`ClusterNotUpError`), it triggers an `up()` to create it.
* **CLI Updates:** Updated the `serve_apply` CLI command to support service arguments and fixed Pylint warnings regarding unused arguments.
* **SDK Compatibility:** Updated the `sky.serve.client.sdk` wrapper to accept the `workers` argument, fixing MyPy type checking errors.


<!-- Describe the tests ran -->

**Details on Manual/New Tests:**
1. **Added Unit Tests:** Created `tests/unit_tests/test_serve_apply.py` which mocks the backend to verify the Upsert logic handles:
    * Cold start (Service not found).
    * Controller down (`ClusterNotUpError`).
    * Hot update (Service exists).
2. **Manual Verification:**
    * Ran `sky serve apply -n test-service service.yaml` on a fresh cluster -> Verified service creation.
    * Ran `sky serve apply -n test-service service.yaml` on an existing service -> Verified service update.
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [X] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [X] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local)
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

<!-- CI commands (/-prefixed) can only be triggered by repo members -->

cblmemo reviewed (commented):
Thanks @jusstinn for contributing to this! Could we reference to managed jobs pool implementation here to align the codepath?

https://github.com/skypilot-org/skypilot/blob/2a5e5fbb5c0ca5469ac637f7920a92c954917dfb/sky/jobs/server/core.py#L1152-L1159

jusstinn committed:
[Serve] Implement server-side 'sky serve apply' API

This PR implements a unified 'sky serve apply' interface. It aligns the codepath with Managed Jobs by moving the Upsert logic (create vs update) to the server side.

Key changes:
- Client: impl.apply now sends a SERVE_APPLY request instead of checking status locally.
- Server: Added /serve/apply endpoint in server.py and core.py to handle the logic.
- Protocol: Added ServeApplyBody to payloads.py.
- CLI: Updated command.py to support the apply arguments.

jusstinn committed:
Fix pylint unused argument warning

jusstinn committed:
Trigger CI retry

jusstinn commented:
Hey @cblmemo ! I modified the code to reference the jobs_pool implementation structure. It now uses a dedicated API endpoint for the apply logic. Awaiting your feedback!

cblmemo commented on sky/server/requests/request_names.py:
```diff
@@ -53,6 +53,7 @@ class RequestName(str, enum.Enum):
     # Serve requests
     SERVE_UP = 'serve.up'
     SERVE_UPDATE = 'serve.update'
+    SERVE_APPLY = 'serve_apply'
```
```suggestion
    SERVE_APPLY = 'serve.apply'
```

jusstinn committed:
Refactor client apply to match job pool and fix naming

cblmemo commented on sky/serve/client/impl.py:
```diff
@@ -119,47 +119,86 @@ def apply(
+                    '/jobs/pool_apply',
+                    json=json.loads(body.model_dump_json()),
+                    timeout=(5, None))
+                return server_common.get_request_id(response)
```
This code path looks really similar to the else branch below. can we reuse some code?

cblmemo commented on sky/serve/client/sdk.py:
```diff
@@ -88,6 +88,39 @@ def update(
+def apply(
+    task: Union['sky.Task', 'sky.Dag'],
+    workers: Optional[int],
+    service_name: str,
```
```suggestion
def apply(
    task: Union['sky.Task', 'sky.Dag'],
    service_name: str,
```
The argument `workers` is a pool-only thing. Could we remove it from this sdk and just pass in `None` to `impl.apply`?

cblmemo commented on sky/serve/server/core.py:
```diff
@@ -65,6 +66,20 @@ def update(task: Optional['sky.Task'],
+    """Apply a config to a service."""
+    # Convert YAML string to Task object
+    kwargs = body.to_kwargs()
+    task = kwargs.pop('task')
```
Why do we need those lines? can we do similar implementation with `sky/jobs/server/core.py` to pass task, service_name etc as arguments instead of body>?

cblmemo reviewed (commented):
Sorry for getting back a little bit late, and thanks for contributing @jusstinn ! It looks very near to ready. Left some discussions ;))
